### PR TITLE
Made Adjacency a shared_ptr<>

### PR DESCRIPTION
**Pull request by koonschi**
This has several benefits. 

First, the Adjacency can be created only once per process and be cached in the Initializer, which was not possible before since the adjacency was returned as an instance and getAdjacency() was called multiple times (at least twice IIRC).

Second, users can define their own adjacencies, which allows them to optimize the getNeighbors() function. It's no longer required to store the entire data in every process, if only a fraction is required.

### CODE REVIEW

**Comment by koonschi** [comment]
And it simplifies some code since move overloads for several constructors are no longer required.

File: src/geometry/regionbasedadjacency.h [hunk 1]
```diff
@@ -34,27 +34,35 @@ class RegionBasedAdjacency
...
+#ifdef LIBGEODECOMP_WITH_CPP14
+    RegionBasedAdjacency(RegionBasedAdjacency &&other) = default;
+    RegionBasedAdjacency &operator=(RegionBasedAdjacency &&other) = default;
+#endif // LIBGEODECOMP_WITH_CPP14
 
     /**
      * Insert a single edge (from, to) to the graph
      */
-    void insert(int from, int to);
+    virtual void insert(int from, int to);
```
Comment (gentryx) [inline src/geometry/regionbasedadjacency.h]: why did you make these virtual?

Comment (koonschi) [inline src/geometry/regionbasedadjacency.h]: To enable users to derive from this class and define their own adjacencies. We could also add an own Adjacency interface, that would probably be cleaner.

Comment (gentryx) [inline src/geometry/regionbasedadjacency.h]: I don't like the idea of the user having to subclass from Adjacency as it requires users to know how our code is working internally (see mail for more details). Please remove these virtual qualifiers.

**Comment by gentryx** [comment]
http://cdn.meme.am/instances/500x/57495448.jpg

**Comment by gentryx** [comment]
Thanks! Good points were made.

**Pull request merged**